Feature/#516376

sugisaka-being opened the pull request:
1章演習問題の解答を作成しました。

sugisaka-being committed:
#516376 [C#テキスト]1章演習問題

gitignoreとChapter1のプロジェクトを追加しました。

sugisaka-being committed:
#516376 [C#テキスト]1章演習問題

VSの設定をしました。

sugisaka-being committed:
#516374 [C#テキスト]1章演習問題

1章演習問題の解答を作成しました。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,36 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
不要なusingは削除しておきましょう。
Visual Studio で確認した際にグレー表示になっているものは使用されていません。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,36 @@
+            　3.で名前空間を別の名前空間に変更したため、1.の解答はコメントとして表示しております。*/
+
+            //3.
+            ReChapter1_1_1.Product wDorayaki = new ReChapter1_1_1.Product(98, "どら焼き", 210);
```
現在のままだとProductクラスを利用する度にReChapter1_1_1を記述する必要があります。
usingディレクティブで ReChapter1_1_1の名前空間を指定してみてください。
そうすれば問題1-1-1の解答についてもコメントアウトしなくても大丈夫になると思います。

また、usingの並び順についても[コーディング規約](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#using)に記載があるので、そちらも確認してみてください。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,36 @@
+
+            //2.
+            int wDorayakiTax = wDorayaki.GetTax();
+            Console.WriteLine(wDorayakiTax);           
```
ローカル変数の`wDorayakiTax`は一度しか使われておらず、かつ特に宣言する必要性も感じられません。
このような場合はメソッドの戻り値を直接`Console.WriteLine()`の引数に指定してしまうとよいです。
例) `Console.WriteLine(wDorayaki.GetTax());`

また、現在の出力では税額が数字として表示されますが、数字だけだと何を表しているのか分かりにくいため、"消費税額は〇〇円です。"と出力されるように修正してみてください。

teshima-being replied:
`Console.WriteLine(wDorayakiTax)` の後ろに余分な半角スペースがありますので、削除しておきましょう。
ちなみに、Visual Studio で「編集」> 「詳細」>「スペースの表示」をすることで、スペースが見える化されます。

また、「Ctrl + K」>「Ctrl + D」の順に押下することで、余計な空白を削除してくれたり、インデントを整えてくれたりします。（「Ctrl + K」>「Ctrl + F」で選択範囲のコード整形も可能です。）

teshima-being commented on Chapter1/Chapter1-1-1/Product.cs:
```diff
@@ -0,0 +1,30 @@
+    internal class Product {
+        public int Code { get; private set; }
+        public string Name { get; private set; }
+        public int Price { get; private set; }
```
クラスやプロパティ、コンストラクタ、メソッドには[ドキュメントコメント](https://learn.microsoft.com/ja-jp/visualstudio/ide/reference/generate-xml-documentation-comments?view=vs-2022)をつけましょう。
そうすることにより、そのクラスやプロパティなどにマウスをホバーした際に、記述したコメントが確認できるようになります。
![スクリーンショット 2024-10-15 110650](https://github.com/user-attachments/assets/82457add-6a3b-4767-a8ba-8ff84c26d63c)
クラス外からプロパティなどを使うときに、そのプロパティが定義されているクラスにわざわざ移動しなくても、どういったプロパティなのかをひと目で確認できるようになるため、非常に便利です。

ちなみに、クラスやプロパティなどのすぐ上の行で///と入力すると、自動で補完してくれます。

teshima-being commented on Chapter1/Chapter1-1-1/Product.cs:
```diff
@@ -0,0 +1,30 @@
+
+        //消費税を求める
+        public int GetTax() {
+            return (int)(Price * 0.08);
```
クラス内のプロパティを参照するときはthisをつけるようにしましょう。
thisについては[コーディング規約](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#:~:text=this%0A%E3%80%80%E3%80%80%E3%83%97%E3%83%AD%E3%83%91%E3%83%86%E3%82%A3%EF%BC%9A%E3%82%AF%E3%83%A9%E3%82%B9%E5%86%85%E3%81%A7%E3%83%97%E3%83%AD%E3%83%91%E3%83%86%E3%82%A3%E3%82%92%E5%8F%82%E7%85%A7%E3%81%99%E3%82%8B%E5%A0%B4%E5%90%88%E3%81%AF%20this.Property%20%E3%81%A8%E6%9B%B8%E3%81%8F%E3%81%AE%E3%82%92%E6%8E%A8%E5%A5%A8%E3%80%82%0A%E3%80%80%E3%80%80%E3%83%95%E3%82%A3%E3%83%BC%E3%83%AB%E3%83%89%EF%BC%9A%E6%8E%A5%E9%A0%AD%E8%BE%9E%E3%81%AE%20%22F%22%20%E3%81%A7%E8%87%AA%E8%BA%AB%E3%81%AE%E3%83%95%E3%82%A3%E3%83%BC%E3%83%AB%E3%83%89%E3%81%A0%E3%81%A8%E5%88%86%E3%81%8B%E3%82%8B%E3%81%AE%E3%81%A7%20this%20%E3%82%92%E4%BB%98%E3%81%91%E3%81%AA%E3%81%8F%E3%81%A6%E3%82%82%E3%81%84%E3%81%84%E3%80%82%E4%BB%98%E3%81%91%E3%81%A6%E3%82%82%E3%81%84%E3%81%84%E3%80%82%0A%E3%80%80%E3%80%80%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%EF%BC%9A%E3%80%80this%20%E3%82%92%E4%BB%98%E3%81%91%E3%81%A6%E3%82%82%E4%BB%98%E3%81%91%E3%81%AA%E3%81%8F%E3%81%A6%E3%82%82%E3%81%84%E3%81%84%E3%80%82)にも記載がありますので、一度確認してみてください。

teshima-being commented on Chapter1/Chapter1-1-1/Product.cs:
```diff
@@ -0,0 +1,30 @@
+            return (int)(Price * 0.08);
+        }
+
+        //税込み価格を求める
```
`//`とコメント本文の間には半角スペースを1つ挟んでください。可読性向上のためです。

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -0,0 +1,44 @@
+                Birthday = new DateTime(2001, 11, 21),
+                Grade = 1,
+                ClassNumber = 1,
+            };
```
せっかくなので、Studentクラスにコンストラクタを実装し、実装したコンストラクタを利用してプロパティに値を設定してみてください。

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -0,0 +1,44 @@
+            };
+
+            //3.
+            Console.WriteLine("Student: ({0},{1},{2},{3})", wStudent.Name, wStudent.Birthday, wStudent.Grade, wStudent.ClassNumber);
```
コンソールへの出力の際には、文字列補間式を使ってみてください。
参考：[C# 文字列の文字列補間](https://csharp.keicode.com/basic/string-interpolation.php)
　　　[[C#] 文字列の書式設定を行う方法](https://zenn.dev/rinc5/articles/8e1e68f56993ce)

他の箇所でも同様に文字列補間式で出力してみましょう。

また、現在の出力は "Student: (Reina,2001/11/21 0:00:00,1,1)" となっており、それぞれの数字が何を示すのか分かりにくいです。
それぞれが何を示しているのか分かるように出力してみてください。
例) "氏名：山田花子、学年：1、組：1、生年月日：2001年10月15日"

teshima-being commented on Chapter1/Chapter1-1-2/Program.cs:
```diff
@@ -0,0 +1,69 @@
+
+        static void Main(string[] args) {
+            //2.
+            MyClass wMyClass1 = new MyClass { X = 10, Y = 20 };　//MyClassのオブジェくトを作成
+            MyStruct wMyStruct1 = new MyStruct { X = 10, Y = 20 }; //MyStructのオブジェクトを作成
```
細かいですが、コードから何をしているかすぐに読み取れる部分についてはコメントは不要です。
コメントがある分読むべき箇所が増えてしまいます。
コードを理解するうえで役に立つ場合にコメントを記述するよう意識するとよいです。
（コードを読む人が「なぜ？」と思うであろう箇所にあらかじめコメントしておくなど）

teshima-being reviewed (commented):
レビューしましたので、ご確認ください。
指摘された事項については、他の箇所でも同様に修正すべき箇所がないか確認して、必要に応じて修正するようお願いします。

sugisaka-being committed:
#516376 [C#テキスト]1章演習問題

不要なusingを削除しました。
ドキュメントコメントを追加しました。
//とコメント本文の間に半角スペースを1つ挟みました。
余分な半角スペースを削除しました。
コードから何をしているかすぐに読み取れる部分について、コメントを削除しました。
Chapter1-1-1のProgram.cs「3.」において、usingディレクティブを利用した解答に修正しました。
Chapter1-1-1のProgram.cs「2.」において、メソッドの戻り値を直接Console.WriteLine()の引数に指定しました。
また、税額を数字のみで表示するのではなく、"消費税額は〇〇円です。"と出力されるように修正を行いました。
Chapter1-1-1のProduct.csにおいて、クラス内のプロパティを参照している部分に「this」を追加しました。
Chapter1-1-3において、Studentクラスにコンストラクタを実装し、それを利用してプロパティに値を設定しました。
Chapter1-1-3のProgram.cs「3.」において、文字列補完式を利用して出力を行いました。

m-hotta-being commented on Chapter1/Chapter1-1-1/Product.cs:
```diff
@@ -0,0 +1,50 @@
+    /// </summary>
+    /// 
+    internal class Product {
+        /// <summary>
+        /// 商品コード
+        /// </summary>
+        public int Code { get; private set; }
+
+        /// <summary>
+        /// 商品名
+        /// </summary>
+        public string Name { get; private set; }
+
+        /// <summary>
+        /// 商品価格（税抜き）
+        /// </summary>
+        public int Price { get; private set; }
```
【FYI: 情報共有。アクション不要です。】
こうしておくことで readonly なプロパティにできます。
コンストラクタで値を設定した後は変更されないことが明らかになるので、読むときに考慮することが減って読みやすくなります。
```suggestion
        /// <summary>
        /// 商品コード
        /// </summary>
        public int Code { get; }

        /// <summary>
        /// 商品名
        /// </summary>
        public string Name { get; }

        /// <summary>
        /// 商品価格（税抜き）
        /// </summary>
        public int Price { get; }
```

teshima-being commented on Chapter1/Chapter1-1-1/Product.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <summary>
+    /// 商品クラス
+    /// </summary>
+    /// 
```
細かいですが、1行余分な行があるので削除しておきましょう。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,29 @@
+        　　ただし、MainメソッドのあるProgramクラスの名前空間はそのままとしてください */
+
+        static void Main(string[] args) {
+            // 1.
+            Product wDorayaki = new Product(98, "どら焼き", 210);
```
※参考（修正は任意）
テキストでまだ出てきていないかもしれませんが、右辺から変数の型が明らかな場合はvar（暗黙的な型指定）を利用して、下記のように記述するとよいです。また時間があれば調べてみてください。
var wDorayaki = new Product(98, "どら焼き", 210);

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,29 @@
+            // 1.
+            Product wDorayaki = new Product(98, "どら焼き", 210);
+
+            // 2.
+            Console.WriteLine("消費税額は{0}円です。", wDorayaki.GetTax());
```
ここも文字列補間式で書き換えてみましょう。
文字列補間式で記述したほうが、左から読み進める際に変数部分が何を示しているのか分かりやすいです。

Chapter1-1-2のProgram.csの出力も文字列補間式に書き換えてみましょう。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,29 @@
+        　　どら焼きオブジェクトの消費税額を求め、コンソールに出力するコードを書いてください。
+
+        　　3.
+        　　Productクラスが属する名前空間を別の名前空間に変更し、Maiaメソッドから呼び出すようにしてください。
+        　　ただし、MainメソッドのあるProgramクラスの名前空間はそのままとしてください */
```
ついでに誤字も修正しておきましょう。
Maia → Main

teshima-being commented on Chapter1/Chapter1-1-2/MyStruct.cs:
```diff
@@ -1,12 +1,16 @@
-using System.Threading.Tasks;
-
-namespace Chapter1_1_2 {
+﻿namespace Chapter1_1_2 {
```
usingディレクティブとnamespaceの間には1行改行がある方が見やすいです。
新規でプロジェクトを作成したときも、ここにはあらかじめ1行空行が入っているかと思います。

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -1,42 +1,36 @@
+            // 2.
+            Student wStudent = new Student(1, 1) {
                 Name = "Reina",
                 Birthday = new DateTime(2001, 11, 21),
```
学年とクラスはコンストラクタでプロパティに値を設定できるようになりましたね！
名前と誕生日についても同じようにコンストラクタでプロパティに値を設定できるようにしてみましょう。

おそらく、StudentクラスにNameプロパティとBirthdayプロパティが記述されていないために別にしたのかと推測しますが、StudentクラスはPersonクラスを継承しているので、どちらのプロパティも利用できます。

teshima-being reviewed (commented):
とても分かりやすくなりました！
あと少しだけ修正をお願いします。

sugisaka-being committed:
#516376 [C#テキスト]1章演習問題

Chapter1-1-1のProduct.csにおいて、余分な行を削除しました。
Chapter1-1-1のProgram.csにおいて、誤字を修正しました。
Chapter1-1-1とChapter1-1-2のProgram.csにおいて、変数の型が明らかな部分にvarを利用しました。
Chapter1-1-1とChapter1-1-2において、出力を文字列補完式に書き換えました。
Chapter1-1-3において、usingディレクティブとnamespaceの間に改行を挿入しました。
Chapter1-1-3のProgram.csにおいて、コンストラクタで（名前と誕生日）プロパティに値を設定しました。

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -22,10 +22,7 @@ internal class Program {
+            var wStudent = new Student("Reina", new DateTime(2001, 11, 21), 1, 1) {};
 
             // 3.
             Console.WriteLine($"氏名：{wStudent.Name}、学年：{wStudent.Grade}、組：{wStudent.ClassNumber}、生年月日：{wStudent.Birthday:yyyy年M月d日})");
```
すみません。前回の指摘漏れです。
`生年月日：{wStudent.Birthday:yyyy年M月d日})"`の最後の`)`が余分ですね。
今のままだと`2001年11月21日)`と出力されます。

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -22,10 +22,7 @@ internal class Program {
-                Name = "Reina",
-                Birthday = new DateTime(2001, 11, 21),
-            };
+            var wStudent = new Student("Reina", new DateTime(2001, 11, 21), 1, 1) {};
```
コンストラクタで各プロパティを設定できるようになり、最後の`{}`は不要になったので、削除しておきましょう。

teshima-being commented on Chapter1/Chapter1-1-3/Person.cs:
```diff
@@ -0,0 +1,31 @@
+        /// <returns>年齢</returns>
+        public int GetAge() {
+            DateTime wToday = DateTime.Today;
+            int wAge = wToday.Year - Birthday.Year;
+            if (wToday < Birthday.AddYears(wAge)) {
```
すみません。ここも前回の指摘漏れです。
`Birthday`に`this`を付けておきましょう。

teshima-being reviewed (commented):
あと少しです！
修正よろしくお願いします。

sugisaka-being committed:
#516376 [C#テキスト]1章演習問題

Chapter1-1-3のProgram.csにおいて、生年月日：{wStudent.Birthday:yyyy年M月d日})"の最後の余分な「)」を削除しました。
Chapter1-1-3のProgram.csにおいて、不要な「{}」を削除しました。
Chapter1-1-3のPerson.csにおいて、Birthdayにthisを付けました。

teshima-being commented:
レビューOKです。1章お疲れ様でした。
